### PR TITLE
Switch strategy to cloudformation

### PR DESCRIPTION
**Pull request by szuecs**
This will change all ALB creation calls to CloudFormation.
We have to review and update IAM policy.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -92,53 +122,62 @@ resource anymore, it deletes all the previously created resources.
...
 ```
 
-To complete the setup, you'll need to fulfill some additional requirements regarding security groups and IAM roles; [more info here](deploy/requirements.md).
+To complete the setup, you'll need to fulfill some additional requirements regarding security groups and IAM 
+roles; [more info here](deploy/requirements.md).
 
 ### DNS
 
-To have convenient DNS names for your application, you can use [External DNS](https://github.com/kubernetes-incubator/external-dns). It's not strictly necessary for this Ingress Controller to work, though.
+To have convenient DNS names for your application, you can use [mate](https://github.com/zalando-incubator/mate).
```
Comment (Raffo) [inline README.md]: Is the README really up to date? this looks like a previous version of it (see the reference to `mate`).

File: aws/adapter.go [hunk 2]
```diff
@@ -159,53 +206,60 @@ func (a *Adapter) PublicSubnetIDs() []string {
...
-		return nil, err
+// GetStack returns the CloudFormation stack details with the name or ID from the argument
+func (a *Adapter) GetStack(stackID string) (*Stack, error) {
+	return getStack(a.cloudformation, stackID)
+}
+
+// DeleteStack deletes the CloudFormation stack with the given name
+func (a *Adapter) DeleteStack(stack *Stack) error {
+	if err := detachTargetGroupFromAutoScalingGroup(a.autoscaling, stack.TargetGroupARN(), a.AutoScalingGroupName()); err != nil {
+		return fmt.Errorf("DeleteStack failed to dettach: %v", err)
```
Comment (Raffo) [inline aws/adapter.go]: Typo: "detach".

File: aws/adapter_test.go [hunk 3]
```diff
@@ -0,0 +1 @@
+package aws
```
Comment (Raffo) [inline aws/adapter_test.go]: This file is empty.

File: cloudformation.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+into your Kubernetes services.
+
+This is achieved by creating an internet-facing Application Load Balancer (ALB) with 2 default listeners 
+(HTTP and HTTPS) and a Target Group that is attached to the Auto Scaling Group of the worker nodes.
+
+You can further customize the stack adding other resources or tags for your specific setup.
+
+## Customization
+
+The controller accepts some flags that to customize some basic behavior of the stack. You can specify the health
```
Comment (Raffo) [inline cloudformation.md]: Please drop "that"

File: cloudformation.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+## Resources
+
+The built-in template creates a set of AWS resources necessary to accept and forward requests from the outside world
+into your Kubernetes services.
+
+This is achieved by creating an internet-facing Application Load Balancer (ALB) with 2 default listeners 
+(HTTP and HTTPS) and a Target Group that is attached to the Auto Scaling Group of the worker nodes.
+
+You can further customize the stack adding other resources or tags for your specific setup.
```
Comment (Raffo) [inline cloudformation.md]: by adding

File: worker.go [hunk 6]
```diff
@@ -2,155 +2,188 @@ package main
...
+	log.Printf("creating stack for certificate %q / ingress %q", certificateARN, item.ingress)
 
-	lb, err := awsAdapter.CreateLoadBalancer(certificateARN)
+	stackId, err := awsAdapter.CreateStack(certificateARN)
 	if err != nil {
-		return nil, fmt.Errorf("failed to create ALB for certificate %q. %v\n", certificateARN, err)
+		if isAlreadyExistsError(err) {
+			item.stack, err = awsAdapter.GetStack(stackId)
+			if err == nil {
+				return
```
Comment (Raffo) [inline worker.go]: What do we want to return here?

File: worker.go [hunk 7]
```diff
@@ -2,155 +2,188 @@ package main
...
+		return awsErr.Code() == cloudformation.ErrCodeAlreadyExistsException
 	}
+	return false
 }
 
-func deleteOrphanedLoadBalancers(awsAdapter *aws.Adapter, ingresses []*kubernetes.Ingress) error {
-	lbs, err := awsAdapter.FindManagedLoadBalancers()
-	if err != nil {
-		return err
+func updateIngress(kubeAdapter *kubernetes.Adapter, item *managedItem) {
```
Comment (Raffo) [inline worker.go]: this function does not return any error. Are we fine with this?

File: worker.go [hunk 8]
```diff
@@ -2,155 +2,188 @@ package main
...
 
-	for _, lb := range lbs {
-		if _, has := certificateMap[lb.CertificateARN()]; !has {
-			if err := awsAdapter.DeleteLoadBalancer(lb); err == nil {
-				log.Printf("deleted orphaned load balancer ARN %q\n", lb.ARN())
-			} else {
-				log.Printf("failed to delete orphaned load balancer ARN %q: %v\n", lb.ARN(), err)
-			}
-		}
+func deleteStack(awsAdapter *aws.Adapter, item *managedItem) {
```
Comment (Raffo) [inline worker.go]: same as above, errors are swallowed. IF this is intended we should at least make a comment to make this clear.

File: aws/adapter.go [hunk 9]
```diff
@@ -159,53 +206,60 @@ func (a *Adapter) PublicSubnetIDs() []string {
...
 		return nil, err
 	}
-	return lbs, nil
+	targetGroupARNs := make([]string, len(stacks))
+	for i, stack := range stacks {
+		targetGroupARNs[i] = stack.targetGroupARN
+	}
+	// This call is idempotent and safe to execute every time
+	if err := attachTargetGroupsToAutoScalingGroup(a.autoscaling, targetGroupARNs, a.AutoScalingGroupName()); err != nil {
+		log.Printf("FindManagedStacks() failed to attach target groups to ASG: %v", err)
```
Comment (Raffo) [inline aws/adapter.go]: we log the error, but we do not return it. is it okay to ignore?

File: aws/adapter.go (hunk 2 above)
Comment (szuecs) [inline aws/adapter.go]: fixed

File: aws/adapter_test.go (hunk 3 above)
Comment (szuecs) [inline aws/adapter_test.go]: removed

File: cloudformation.md (hunk 5 above)
Comment (szuecs) [inline cloudformation.md]: fixed

File: cloudformation.md (hunk 4 above)
Comment (szuecs) [inline cloudformation.md]: fixed

File: worker.go (hunk 6 above)
Comment (szuecs) [inline worker.go]: This is implicit: stack = nil, if it's not ready and later it is checked in updateIngress() to return fast (doWork() for-switch statement) and being retriggered in the next loop.

File: worker.go (hunk 7 above)
Comment (szuecs) [inline worker.go]: The comment is done in the isComplete() comment

File: worker.go (hunk 8 above)
Comment (szuecs) [inline worker.go]: The comment is done in the isComplete() comment

File: aws/adapter.go (hunk 9 above)
Comment (szuecs) [inline aws/adapter.go]: Yes.
The comment is done in the isComplete() comment

File: README.md (hunk 1 above)
Comment (szuecs) [inline README.md]: fixed the README, thanks for this catch!

**Comment by Raffo** [comment]
👍

**Comment by szuecs** [comment]
:+1:

**Pull request merged**